File: apps/desktop/src/lib/control/editorFileWatch.ts
import { Channel, invoke } from "@tauri-apps/api/core";

export type EditorFileWatchEvent =
  | {
      kind: "changed";
      subscription_id: string;
      project_id: string;
      resource_id: string;
      paths: string[];
    }
  | {
      kind: "scm_changed";
      subscription_id: string;
      project_id: string;
      resource_id: string;
    }
  | {
      kind: "failed";
      subscription_id: string;
      project_id: string;
      message: string;
    };

export async function watchEditorFiles(
  projectId: string,
  resourceIds: string[],
  onEvent: (event: EditorFileWatchEvent) => void,
): Promise<() => Promise<void>> {
  const channel = new Channel<EditorFileWatchEvent>();
  channel.onmessage = onEvent;
  const subscriptionId = await invoke<string>("editor_file_watch_start", {
    projectId,
    resourceIds,
    onEvent: channel,
  });
  let stopped = false;
  return async () => {
    if (stopped) return;
    stopped = true;
    await invoke("editor_file_watch_stop", { subscriptionId });
  };
}

File: apps/desktop/src-tauri/src/editor_file_watch.rs
use std::sync::Arc;

use nucleus_server::EditorFileWatchEventSink;
use tauri::ipc::Channel;

use crate::DesktopState;

#[tauri::command]
pub fn editor_file_watch_start(
    state: tauri::State<'_, DesktopState>,
    project_id: String,
    resource_ids: Vec<String>,
    on_event: Channel<nucleus_server::EditorFileWatchEvent>,
) -> Result<String, String> {
    let sink: EditorFileWatchEventSink = Arc::new(move |event| {
        let _ = on_event.send(event);
    });
    state
        .editor_file_watch
        .start(&state.server_state, &project_id, &resource_ids, sink)
}

#[tauri::command]
pub fn editor_file_watch_stop(
    state: tauri::State<'_, DesktopState>,
    subscription_id: String,
) -> Result<(), String> {
    state.editor_file_watch.stop(&subscription_id)
}
